spyder_kernels/customize/spydercustomize.py

```diff
@@ -1007,26 +1052,25 @@ def runcell(cellname, filename=None):
         # Nothing to execute
         return
 
-    namespace = _get_globals()
-    # Save previous __file__
-    namespace_has_file = '__file__' in namespace
-    if namespace_has_file:
-        namespace_filename = namespace['__file__']
-    namespace['__file__'] = filename
-
     # Trigger `post_execute` to exit the additional pre-execution.
     # See Spyder PR #7310.
     ipython_shell.events.trigger('post_execute')
 
-    if PY2:
-        filename = maybe_encode(filename)
-    exec(compile(cell_code, filename, 'exec'), namespace)
-
-    # Restore __file__
-    if namespace_has_file:
-        namespace['__file__'] = namespace_filename
-    else:
-        namespace.pop('__file__')
+    with FileNamespace(filename, current_namespace=True) as namespace:
+        try:
+            if PY2:
+                filename = maybe_encode(filename)
+            exec(compile(cell_code, filename, 'exec'), namespace)
```

Comment:
If we don't use IPython's `run_cell`, does that mean that IPython magics still work?

Reply:
Do we expect them to work? I mean if I run a file they don't work, wo why should they work when running a cell?

Reply:
In my opinion, cells are just file fragments

Reply:
It would be nice. We are trying to make it so you can run a file as cells like a Jupyter notebook.

Reply:
Cell magic current works for cells in spyder3

Reply:
It looks like vscode can runcells with magics

Reply:
I don't think we should allow magics in Python files because they're not valid code.